Feature/create update delete sort

dhruv-d-simform opened the pull request:
# Description

Added following features : 
- Create User
- Update User
- Delete User
- Sort Users

# Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Refractoring
- [ ] Performance improvement
- [ ] Scaffolding

# Screenshots

![Screenshot from 2025-04-09 13-45-57](https://github.com/user-attachments/assets/27f80613-8197-4fc3-925d-2acdf56a944e)
![Screenshot from 2025-04-09 13-46-12](https://github.com/user-attachments/assets/01f43cc2-06e7-4c62-8950-31d8d5c9e2de)

# Validations

![Screenshot from 2025-03-31 15-07-34](https://github.com/user-attachments/assets/a780d0dd-3a03-465a-9c86-4a8a9a06c659)
> [!NOTE] 
> Checkout this issue https://github.com/dhruvdhaduk-simform/user-management-ts/issues/3
> ![Screenshot from 2025-03-31 15-08-06](https://github.com/user-attachments/assets/a682b8e1-cd6d-459b-8042-8550120f8924)   

![Screenshot from 2025-03-31 15-08-43](https://github.com/user-attachments/assets/a99a3581-0dd2-4eec-9525-8eb18bb0b315)
![Screenshot from 2025-03-31 15-08-50](https://github.com/user-attachments/assets/55ab67a9-6288-44ac-b064-9a397fcb8cdc)


# How to test

Deployment preview : https://deployment-preview-5-user-management.netlify.app/

dhruv-d-simform committed:
feature: Add createUser method in Store class

dhruv-d-simform committed:
feature: Added submit handler in View class for Add User form

dhruv-d-simform committed:
feature: Cleanup the index.ts and pass selector for add user form in View

dhruv-d-simform committed:
docs: Add comments in View class for submit handler

dhruv-d-simform committed:
feature: Add delete user logic in Store class

dhruv-d-simform committed:
feature: Attach the Delete user functionality to UI

dhruv-d-simform committed:
feature: Add method to Update User in Store class

dhruv-d-simform committed:
feature: Add event handlers to edit buttons

dhruv-d-simform committed:
feature: Create sort method for Store class

dhruv-d-simform committed:
feature: Attach sort handlers to UI

krushna-simform commented:
Ui looks Good!

But, it is accepting empty string and negative values.

![Screenshot from 2025-04-01 10-29-40](https://github.com/user-attachments/assets/354dce36-b132-4b22-a29e-47e35aa5edaa)

dhruv-d-simform committed:
fix: Prevent user from entering empty strings and negative values in Add user form

dhruv-d-simform commented:
> Ui looks Good!
> 
> But, it is accepting empty string and negative values.
> 
> ![Screenshot from 2025-04-01 10-29-40](https://private-user-images.githubusercontent.com/194056780/428881889-354dce36-b132-4b22-a29e-47e35aa5edaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0ODQwODksIm5iZiI6MTc0MzQ4Mzc4OSwicGF0aCI6Ii8xOTQwNTY3ODAvNDI4ODgxODg5LTM1NGRjZTM2LWIxMzItNGIyMi1hMjllLTQ3ZTM1YWE1ZWRhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMVQwNTAzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZWU3Njk2ZDBmYzQ5Y2U5NzIxODhjZmZjOTVlZGE0OTdhYzAzYTllMjZjYmNlMTFlOGU4MTU0MGFjNjQzNzE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZlbfVGImHrpJBZKYz_BlDGGw48U4lk4qUYTnVlJRxsI)

Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/bdcc1d1dec4855b1ba2a3bd73c46d06cfac600c3

krushna-simform commented:
> > Ui looks Good!
> > But, it is accepting empty string and negative values.
> > ![Screenshot from 2025-04-01 10-29-40](https://private-user-images.githubusercontent.com/194056780/428881889-354dce36-b132-4b22-a29e-47e35aa5edaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0ODQwODksIm5iZiI6MTc0MzQ4Mzc4OSwicGF0aCI6Ii8xOTQwNTY3ODAvNDI4ODgxODg5LTM1NGRjZTM2LWIxMzItNGIyMi1hMjllLTQ3ZTM1YWE1ZWRhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMVQwNTAzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZWU3Njk2ZDBmYzQ5Y2U5NzIxODhjZmZjOTVlZGE0OTdhYzAzYTllMjZjYmNlMTFlOGU4MTU0MGFjNjQzNzE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZlbfVGImHrpJBZKYz_BlDGGw48U4lk4qUYTnVlJRxsI)
> 
> Fixed it. [bdcc1d1](https://github.com/dhruvdhaduk-simform/user-management-ts/commit/bdcc1d1dec4855b1ba2a3bd73c46d06cfac600c3)

Can we also handle empty string value and negative value for edit mode.
![Screenshot from 2025-04-01 13-00-03](https://github.com/user-attachments/assets/02dc1ca2-7fd7-4ae5-8177-dc2531307a26)

krushna-simform commented:
Can we add one `alert` message when user press `delete` button ?

like `Are you sure you want to delete this user?`

dhruv-d-simform committed:
fix: Validate data before updating user

dhruv-d-simform committed:
feature: Ask for confirmation before deleting user

dhruv-d-simform commented:
> 
> Can we also handle empty string value and negative value for edit mode. ![Screenshot from 2025-04-01 13-00-03](https://private-user-images.githubusercontent.com/194056780/428920283-02dc1ca2-7fd7-4ae5-8177-dc2531307a26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0OTMwNzIsIm5iZiI6MTc0MzQ5Mjc3MiwicGF0aCI6Ii8xOTQwNTY3ODAvNDI4OTIwMjgzLTAyZGMxY2EyLTdmZDctNGFlNS04MTc3LWRjMjUzMTMwN2EyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMVQwNzMyNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTczNjhkMjg2NTJlMjgyOTIwM2RhYmU1NWNmYzc2OTc3MjU3YTViYzMxNGFmMTFmNmVmY2E1NjI4ZDA0OTc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.XqjEd7gWQY-mXxGOCnhWrPccTSSh5XUexBSUMq_JJ6U)

Added validations.

https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/3acea662f030c5a29b44c16d029b2047d82b3623

dhruv-d-simform commented:
> Can we add one `alert` message when user press `delete` button ?
> 
> like `Are you sure you want to delete this user?`

Added alert before deleting.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/1a95eeb740f6c4de9ad2509cfaa9bc7d1850d796

dhruv-d-simform committed:
Add validation for email in edit user operation

krushna-simform commented:
Can we handle this case also for email and phone number.

![Screenshot from 2025-04-01 13-44-51](https://github.com/user-attachments/assets/8ba95d6c-3594-499c-8663-14f383544854)

dhruv-d-simform committed:
fix: Add validation for phone number in Edit user

dhruv-d-simform commented:
> Can we handle this case also for email and phone number.
> 
> ![Screenshot from 2025-04-01 13-44-51](https://private-user-images.githubusercontent.com/194056780/428935756-8ba95d6c-3594-499c-8663-14f383544854.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0OTU3NTcsIm5iZiI6MTc0MzQ5NTQ1NywicGF0aCI6Ii8xOTQwNTY3ODAvNDI4OTM1NzU2LThiYTk1ZDZjLTM1OTQtNDk5Yy04NjYzLTE0ZjM4MzU0NDg1NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMVQwODE3MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NjBmNWUzNTQzOTk0N2ZiMjZmNGI2NWNjZDcwMTYxNzc2ZWZlZWUwNzM4N2VmYjk1ZDA1ZjgyOGQwMTQ1NDEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.gCDo3RNdzCxh2MzUMl1PU3lsieSs-jIMfncVV2_DHuI)

Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/c1f1bf22dec810b5ddcab354d28f46ea5d29b119
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/870dc731c28a0c116b03d459ffa871aebbcac020

krushna-simform commented on src/scripts/store.ts:
```diff
@@ -23,39 +30,69 @@ export default class Store {
         });
     }
 
+    getDeletedUsers(): number[] {
```
One minor suggestion. 

```suggestion
    getDeletedUsers(): Array<number> {
```

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/ee7c8b7ba67d5ec005f2a3dc03b4ec922196176e

dhruv-d-simform committed:
refractor: Use Array<number> for number[]

Sakshi-Simform reviewed (approved):
ui looks good

dhruv-d-simform committed:
refractor: Remove unnecessary commented line

Kaushal-SF commented:
@dhruvdhaduk-simform Few observations from UI:

1) We don't have any validation for first and last name. I can add only numbers value and it is not complaining at all.
<img width="618" alt="image" src="https://github.com/user-attachments/assets/3522ee11-29a4-45f4-98af-8573590ac5a8" />

Kaushal-SF commented:
@dhruvdhaduk-simform For the phone number I can add combination of string and number while adding a user. Which is not valid. We can add limit to the input field as well.

<img width="208" alt="image" src="https://github.com/user-attachments/assets/0efbca44-a428-4ed2-9b0a-b42f0cc02d7d" />

Kaushal-SF commented:
@dhruvdhaduk-simform When I do `Edit`, there is no visual indication that how user can edit the fields. User might have a hard time to figure how the values can be edit.

<img width="814" alt="image" src="https://github.com/user-attachments/assets/f86706c4-df7a-425b-9b9f-099892700b28" />

Kaushal-SF commented:
@dhruvdhaduk-simform Overall UI is looking average. As per your potential, I think you have created a much better UI than this. I would like to know your thoughts on this. Just dark theme and usage of `blue` and `red` buttons seem safer choice for the UI.

Kaushal-SF commented:
@dhruvdhaduk-simform While editing a gender, we are only supporting two options. `Male` and `Female`. However, we are allow to write any string value but when we try to save it, it will not take the added value. IMO, this is not a good UX.

One simple fix, would be to convert it in the light mode and use a bit elegant colors.

dhruv-d-simform committed:
refractor: Create a seperate helper function for attaching edit event listener

dhruv-d-simform committed:
feature: Focus on first name when edit mode is started

dhruv-d-simform committed:
feature: Use form for update users instead of contentEditable

dhruv-d-simform committed:
refractor: Change the variable name from addUserForm to userForm in view.ts

dhruv-d-simform committed:
feature: Change the heading of user form depending on add/update operation

dhruv-d-simform committed:
feature: Change the theme to light mode

dhruv-d-simform committed:
feature: Add validations for firstName, lastName, phone inputs to prevent users from entering invalid characters

dhruv-d-simform committed:
feature: Add character limit to form input fields

dhruv-d-simform committed:
feature: Set the form status to Add User mode when clicked on Add User button

dhruv-d-simform commented:
> @dhruvdhaduk-simform Few observations from UI:
> 
> 1. We don't have any validation for first and last name. I can add only numbers value and it is not complaining at all.
> 
> <img alt="image" width="618" src="https://private-user-images.githubusercontent.com/97082607/430776503-3522ee11-29a4-45f4-98af-8573590ac5a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMTQ2NDIsIm5iZiI6MTc0NDAxNDM0MiwicGF0aCI6Ii85NzA4MjYwNy80MzA3NzY1MDMtMzUyMmVlMTEtMjlhNC00NWY0LTk4YWYtODU3MzU5MGFjNWE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDA4MjU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiNjFkYTlhMzM2ZjcyNTk1MWNmNGI3OTNiNTcwNjE1MDc2MGVmZDljODg4ZTdmOTU5YzMyMTY4ZjU5YzE1NTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7pgmR09xWvYCKR8NVrdLBWeCynz1GbZ0TD_FVNAWMaw">

Added validations as well as input limit.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/645045d31d3f94864e68bb6f1cf126854a76130a
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/a2944784578569bcbe83c98239fcc3130db156e8

dhruv-d-simform commented:
> @dhruvdhaduk-simform For the phone number I can add combination of string and number while adding a user. Which is not valid. We can add limit to the input field as well.
> 
> <img alt="image" width="208" src="https://private-user-images.githubusercontent.com/97082607/430776709-0efbca44-a428-4ed2-9b0a-b42f0cc02d7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMTQ2NDIsIm5iZiI6MTc0NDAxNDM0MiwicGF0aCI6Ii85NzA4MjYwNy80MzA3NzY3MDktMGVmYmNhNDQtYTQyOC00ZWQyLTliMGEtYjQyZjBjYzAyZDdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDA4MjU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwZDBkNWY2OWUxYjVmYmZjZDgxNGFmYmRkODc3ZDkyYjRjNDUyMmQ1NDI0MWUxZmZhOGEyOTNjYjU4Mjg5MzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dbSpk73A9cuUfE364iDEmGatSZNgmEltKBVJH8R5Vf4">

Added validations as well as input limit.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/645045d31d3f94864e68bb6f1cf126854a76130a
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/a2944784578569bcbe83c98239fcc3130db156e8

dhruv-d-simform commented:
> @dhruvdhaduk-simform When I do `Edit`, there is no visual indication that how user can edit the fields. User might have a hard time to figure how the values can be edit.
> 
> <img alt="image" width="814" src="https://private-user-images.githubusercontent.com/97082607/430777521-f86706c4-df7a-425b-9b9f-099892700b28.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMTQ2NDIsIm5iZiI6MTc0NDAxNDM0MiwicGF0aCI6Ii85NzA4MjYwNy80MzA3Nzc1MjEtZjg2NzA2YzQtZGY3YS00MjViLTliOWYtMDk5ODkyNzAwYjI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDA4MjU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2ZjEwOTkxOWJjNjVhYzY5NmMwM2NhYWE3ODUwYzRmMDg3NDZkYWVmYTk4NzYyMWJkYWU1YzM0MDQ1ZjNhZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uCudkS8xWqlCzB_QRAsBqMaFv7Bw_5Sh0wfRgEXs7yk">

Removed this inline editing altogether. Used the popup form to edit users.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/e9d2f25b8ac6cdbe5f1a9d5cb28307507e16c2cb

dhruv-d-simform commented:
> @dhruvdhaduk-simform Overall UI is looking average. As per your potential, I think you have created a much better UI than this. I would like to know your thoughts on this. Just dark theme and usage of `blue` and `red` buttons seem safer choice for the UI.

Used the light theme as you suggested.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/8b0aaf78bbbbd79f7edb504c258b3054af15df28
Let me know if you have any other suggestions.

dhruv-d-simform commented:
> @dhruvdhaduk-simform While editing a gender, we are only supporting two options. `Male` and `Female`. However, we are allow to write any string value but when we try to save it, it will not take the added value. IMO, this is not a good UX.
> 
> One simple fix, would be to convert it in the light mode and use a bit elegant colors.

Fixed it by using form for editing users.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5#issuecomment-2782442798

Kaushal-SF commented on src/index.html:
```diff
@@ -136,6 +140,7 @@ <h2>Add User</h2>
                             name="phone"
                             type="tel"
                             placeholder="Enter Phone Number"
+                            maxlength="20"
```
Let's keep the max length to 10. I think there will be nobody who has phone number of length 20 😅.

dhruv-d-simform replied:
Yes, I know that phone numbers have 10 digits. But I am allowing more characters to handle this format :
`+49 614-958-9364`

Kaushal-SF replied:
In that case, keep max length to `15`. That's a maximum length for any phone number (got this number on the internet)

Kaushal-SF replied:
Also, there could be a minimum length as well. Please set up a minimum length as well. Provide an error if user not provide proper input.

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/48ad494e58e404614a5730a509aba369075bb57d

Kaushal-SF commented on src/index.html:
```diff
@@ -71,31 +71,33 @@ <h1>User Management System</h1>
                             name="first-name"
                             type="text"
                             placeholder="Enter First Name"
+                            maxlength="40"
```
We can allow the user to have combination of string and number for First and Last name.

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/4247130fd2e86b2f5d21d22b164227be6cd34f4e

Kaushal-SF commented on src/index.html:
```diff
@@ -125,6 +128,7 @@ <h2>Add User</h2>
                             id="email"
                             name="email"
                             type="email"
+                            maxlength="254"
```
Isn't it much bigger length for email address?

dhruv-d-simform replied:
I think it is the official max. length for email addresses. Correct me if I am wrong.

Kaushal-SF replied:
Oh, I didn't knew about it.

Kaushal-SF commented:
@dhruvdhaduk-simform It's a good idea to provide a loader while data is loading. Instead of just showing the table columns.

<img width="1458" alt="image" src="https://github.com/user-attachments/assets/b87f6a2e-ac7d-4123-8eea-df36a44e2e52" />

Kaushal-SF commented:
Noticed one thing that, when the pop-up is open, I can still access the `Edit` or `Delete` buttons. And it is allowing me to change the values in the current form. We should avoid it.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/4cac867d-350f-45cb-8724-6dfede944aa9" />

Kaushal-SF commented:
@dhruvdhaduk-simform Can we add inline error for the input fields? It will make more sense rather then checking everything at the end. You can check Krushna's UI for inspiration.

Kaushal-SF commented on src/scripts/store.ts:
```diff
@@ -77,4 +114,68 @@ export default class Store {
+        }
+    }
+
+    sort(attribute: 'firstName' | 'lastName' | 'age', order: 'asc' | 'desc') {
```
Can extract these two types in the separate types. It will make it easier to add or remove fields in the future.

dhruv-d-simform replied:
Created seperate types.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/325b58608e688f0e8c7637005c8c2be045406080

Kaushal-SF commented on src/scripts/view.ts:
```diff
@@ -22,8 +33,195 @@ export default class View {
+        const genderInput = this.#userForm['gender'].value;
+
+        // Validate the form data.
+        if (typeof firstNameInput !== 'string') {
+            alert('Invalid First Name');
+            return;
+        }
+        if (typeof lastNameInput !== 'string') {
+            alert('Invlid Last Name');
+            return;
+        }
```
I am not getting these alert while entering the wrong values.

dhruv-d-simform replied:
Can you please elaborate on how are you entering wrong values ?

Kaushal-SF reviewed (commented):
@dhruvdhaduk-simform UI is looking much better now with the light theme.

Added a few comments. However, I didn't review all the code.

Kaushal-SF commented:
@dhruvdhaduk-simform Verified the code. It looks good other then the provided comments.

dhruv-d-simform committed:
fix: Disable pointer events while popover is open

dhruv-d-simform committed:
feature: Add a loading spinner for better user experience

dhruv-d-simform committed:
feature: Allow users to add combination of letters & numbers for first & last name

dhruv-d-simform committed:
refractor: Create seperate types for Sort Attribute and Sort Order

dhruv-d-simform commented:
> @dhruvdhaduk-simform It's a good idea to provide a loader while data is loading. Instead of just showing the table columns.
> 
> <img alt="image" width="1458" src="https://private-user-images.githubusercontent.com/97082607/431030568-b87f6a2e-ac7d-4123-8eea-df36a44e2e52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwODg0NDEsIm5iZiI6MTc0NDA4ODE0MSwicGF0aCI6Ii85NzA4MjYwNy80MzEwMzA1NjgtYjg3ZjZhMmUtYWM3ZC00MTIzLThlZWEtZGYzNmE0NGUyZTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDA0NTU0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiYTc0NjEyNmNkNWQ2NDhmMWQwZjExMDhjYzBjOTc4ZWNiZjIzYzMwMzY1Zjc0ZGMwNjNlODEwYzQ0NDZhM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FYbpzj-6eT_2bp9Ne10qWFHCpxzVCYiOyRT12syjqwo">

Added loading spinner.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/28a735cd73a44c5c9096adf139742cf5874e4168

dhruv-d-simform commented:
> Noticed one thing that, when the pop-up is open, I can still access the `Edit` or `Delete` buttons. And it is allowing me to change the values in the current form. We should avoid it.
> 
> <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/97082607/431031773-4cac867d-350f-45cb-8724-6dfede944aa9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwODg0NDEsIm5iZiI6MTc0NDA4ODE0MSwicGF0aCI6Ii85NzA4MjYwNy80MzEwMzE3NzMtNGNhYzg2N2QtMzUwZi00NWNiLTg3MjQtNmRmZWRlOTQ0YWE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDA0NTU0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0MWU4Zjg2OGQ4OTg1MTU3NDFkMDI5M2Y3M2I3MzlhYWYwZDc2OGVkYTk3ZTlhOWExNmNiMzIyYWNlNDI3YzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P6Tff7kaW8pS6BnAT7-7YxirSv8Z9S1wt2ubJzCUtkQ">

Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/ac88651e58836d74440a6353ff8d49a36389f178

dhruv-d-simform commented:
> @dhruvdhaduk-simform Can we add inline error for the input fields? It will make more sense rather then checking everything at the end. You can check Krushna's UI for inspiration.

I am already checking inputs before user submits the form. I am preventing user from entering anything invalid in the first place. Therefore there will be no error to show, because users cannot add anything invalid.

krushna-simform commented:
Can we add validation for gmail field.

![Screenshot from 2025-04-08 16-27-29](https://github.com/user-attachments/assets/1a33346d-6135-4142-b909-e2b376829aac)

dhruv-d-simform committed:
feature: Add validation for email input

dhruv-d-simform commented:
> Can we add validation for gmail field.
> 
> ![Screenshot from 2025-04-08 16-27-29](https://private-user-images.githubusercontent.com/194056780/431339331-1a33346d-6135-4142-b909-e2b376829aac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTEyNDEsIm5iZiI6MTc0NDExMDk0MSwicGF0aCI6Ii8xOTQwNTY3ODAvNDMxMzM5MzMxLTFhMzMzNDZkLTYxMzUtNDE0Mi1iOTA5LWUyYjM3NjgyOWFhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwOFQxMTE1NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNGFhY2EyYmFhYjU2ODAzNjM0M2ZjYzUwNWU0ZTYyOWU3ZjhkYzRiNTM2NjZlZDdkNTFmZDZhZmMxYWNiNGMwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CDlne08RSEFK6ZdHImHMhZMLxTrm1jJIinb9wL-BD4w)

Added validation for email.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/e326c75914885cdc432e3cfbe383dce44dd615b2

krushna-simform commented on src/scripts/view.ts:
```diff
@@ -87,11 +294,41 @@ export default class View {
+        if (!users.length) return;
+
         // Array to hold rows corresponding to each User.
         const rowsToRender: HTMLTableRowElement[] = [];
```
```suggestion
        const rowsToRender: Array<HTMLTableRowElement>= [];
```

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/3bcb584fa20345c2b2e6df72f70d87fcb3604381

dhruv-d-simform committed:
refractor: Use Array<HTMLTableRowElement> for HTMLTableRowElement[] in view.ts

Kaushal-SF commented:
> > @dhruvdhaduk-simform Can we add inline error for the input fields? It will make more sense rather then checking everything at the end. You can check Krushna's UI for inspiration.
> 
> I am already checking inputs before user submits the form. I am preventing user from entering anything invalid in the first place. Therefore there will be no error to show, because users cannot add anything invalid.

@dhruvdhaduk-simform Let's discuss this more.

dhruv-d-simform committed:
feature: Change the maxlength of phone number input to 15 and added validation for minlength

dhruv-d-simform committed:
feature: Create utility to validate form inputs

dhruv-d-simform committed:
feature: Create UI to show inline error in form

dhruv-d-simform committed:
feature: Used Form Validator utility to show inline errors in form

dhruv-d-simform committed:
[Empty commit - to troubleshoot netlify builds]

dhruv-d-simform commented:
@Kaushal-SF 

I have added inline error for form validation.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/9c3ecef85a57e73fb3de0e33d529dda86ac6b868
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/9bbd3ac74dbc79560243c695de8a0a5be58d997a
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/3b0452e9a5a6880526dec4c6f2fbc6c474f56ac6

And there some issue with Netlify. It was not building the preview for my latest commit in this branch.  
Therefore, I have created another deployment for this branch. Please use this link below to check it UI.  
https://deployment-preview-5-user-management.netlify.app/
(Also added this link in Description of this PR)

Kaushal-SF commented:
@dhruvdhaduk-simform Can keep the `First Name` and `Last Name` instead of `firstName` and `lastName`.

We are able to provide the string to a `age` input field.

<img width="230" alt="image" src="https://github.com/user-attachments/assets/eec2224b-6fad-4e17-aa5b-55587839c5ca" />

Kaushal-SF reviewed (approved):
Provided a one comment. Otherwise code looks good.

dhruv-d-simform committed:
fix: Use 'First Name' & 'Last Name' in inline error instead of camelCase

dhruv-d-simform commented:
> @dhruvdhaduk-simform Can keep the `First Name` and `Last Name` instead of `firstName` and `lastName`.
> 
> We are able to provide the string to a `age` input field.
> 
> <img alt="image" width="230" src="https://private-user-images.githubusercontent.com/97082607/431806348-eec2224b-6fad-4e17-aa5b-55587839c5ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxOTY3NDksIm5iZiI6MTc0NDE5NjQ0OSwicGF0aCI6Ii85NzA4MjYwNy80MzE4MDYzNDgtZWVjMjIyNGItNmZhZC00ZTE3LWFhNWItNTU1ODc4MzljNWNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA5VDExMDA0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzZmFlMGE4ODAwZmEzNTc0MzY0NjUyN2FmNTZlNmNkYjBjMWQ3NTk0NTRhODVlZTEzYzRjMjdjNGJjYTY5NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-RzfeiI84LDMSTYjUraGzYx18JPRVQ18eUYP_nz1jPw">

Fixed the `First Name` and `Last Name`.
https://github.com/dhruvdhaduk-simform/user-management-ts/pull/5/commits/525525d2fdf6079833b70222eef2c171c46e3db2


Regarding the age input : 
I think this is the Firefox issue. If you use Chromium based browser, then it will not allow us to enter letters.

Kaushal-SF commented:
> Regarding the age input : I think this is the Firefox issue. If you use Chromium based browser, then it will not allow us to enter letters.

It's a Safari.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/ea8661e2-743c-4ca5-b3d2-859a283de2c1" />

dhruv-d-simform commented:
> > Regarding the age input : I think this is the Firefox issue. If you use Chromium based browser, then it will not allow us to enter letters.
> 
> It's a Safari.
> 
> <img alt="image" width="500" src="https://private-user-images.githubusercontent.com/97082607/431856631-ea8661e2-743c-4ca5-b3d2-859a283de2c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQyMDQ1OTgsIm5iZiI6MTc0NDIwNDI5OCwicGF0aCI6Ii85NzA4MjYwNy80MzE4NTY2MzEtZWE4NjYxZTItNzQzYy00Y2E1LWIzZDItODU5YTI4M2RlMmMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA5VDEzMTEzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyZmEyOTBlYjNhNTU0YWFhOGEzYmM3NTJjYjFiNTZlNjRhNGUzODc1OTYyMWE0Mzg5YTAwM2UxMTM3ZjliYjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XN6YdE0HujywXzAQY8QPQAmJa_TVJ4NlvMiqAgURRvk">

Then it will have even more weird behaviours. I can't test anything for Safari as I am on linux. It is our luck that they are even allowing us to show this popup properly. 😅

Kaushal-SF commented:
> Then it will have even more weird behaviours. I can't test anything for Safari as I am on linux.

@dhruvdhaduk-simform Please see, if you can find a reason behind this behavior. However, don't invest much time.

dhruv-d-simform commented:
> > Then it will have even more weird behaviours. I can't test anything for Safari as I am on linux.
> 
> @dhruvdhaduk-simform Please see, if you can find a reason behind this behavior. However, don't invest much time.

Yes. I will see if I can find any reason. And in our case, we have already added implemented inline errors, so it will not negatively impact this site.  
Therefore, I am merging this PR now.
